include/deal.II/matrix_free/hanging_nodes_internal.h

```diff
@@ -33,6 +33,93 @@ namespace internal
 {
   namespace MatrixFreeFunctions
   {
+    /**
+     * Here is the system for how we store constraint types in a binary mask.
+     * This is not a complete contradiction-free system, i.e., there are
+     * invalid states that we just assume that we never get.
+     *
+     * If the mask is zero, there are no constraints. Then, there are three
+     * different fields with one bit per dimension. The first field determines
+     * the type, or the position of an element along each direction. The
+     * second field determines if there is a constrained face with that
+     * direction as normal. The last field determines if there is a
+     * constrained edge of a given pair of coordinate planes, but where
+     * neither of the corresponding faces are constrained (only valid in 3D).
+     *
+     * The element is placed in the 'first position' along *-axis. These also
+     * determine which face is constrained. For example, in 2D, if
+     * face_x and type are set, then x = 0 is constrained.
+     */
+    enum ConstraintTypes : unsigned short
+    {
+      unconstrained = 0,
+
+      type_x = 1 << 0,
+      type_y = 1 << 1,
+      type_z = 1 << 2,
+
+      // Element has as a constraint at * = 0 or * = fe_degree face
+      face_x = 1 << 3,
+      face_y = 1 << 4,
+      face_z = 1 << 5,
+
+      // Element has as a constraint at * = 0 or * = fe_degree edge
+      edge_xy = 1 << 6,
+      edge_yz = 1 << 7,
+      edge_zx = 1 << 8
+    };
+
+
+
+    /**
+     * Global operator which returns an object in which all bits are set which
+     * are either set in the first or the second argument. This operator exists
+     * since if it did not then the result of the bit-or <tt>operator |</tt>
+     * would be an integer which would in turn trigger a compiler warning when
+     * we tried to assign it to an object of type UpdateFlags.
+     */
+    inline ConstraintTypes
+    operator|(const ConstraintTypes f1, const ConstraintTypes f2)
+    {
+      return static_cast<ConstraintTypes>(static_cast<unsigned short>(f1) |
+                                          static_cast<unsigned short>(f2));
+    }
+
+
+
+    /**
+     * Global operator which sets the bits from the second argument also in the
+     * first one.
+     */
+    inline ConstraintTypes &
+    operator|=(ConstraintTypes &f1, const ConstraintTypes f2)
+    {
+      f1 = f1 | f2;
+      return f1;
+    }
+
+
+
+#ifdef DEAL_II_COMPILER_CUDA_AWARE
+    __device__ inline ConstraintTypes
+    operator|(const ConstraintTypes f1, const ConstraintTypes f2)
```

Comment:
This is wrong `__device__` is not part of the signature of the function. For `nvcc` this signature is exactly the same as the one in line 82. We have a macro for this use case `DEAL_II_CUDA_HOST_DEV`. It expands to nothing if we don't support cuda and to `__host__ __device__` otherwise.

Reply:
```
/home/runner/work/dealii/dealii/include/deal.II/matrix_free/hanging_nodes_internal.h(105): error: function "dealii::internal::MatrixFreeFunctions::operator|" has already been defined

/home/runner/work/dealii/dealii/include/deal.II/matrix_free/hanging_nodes_internal.h(114): error: function "dealii::internal::MatrixFreeFunctions::operator|=" has already been defined
```